File: app/models/orientation.rb
class Orientation < ActiveHash::Base
    include ActiveHash::Enum
    enum_accessor :id
    self.data = [
        {
            id: 'north',
            rotation: 0,
            dx: 0,
            dy: -1
        },
        {
            id: 'south',
            rotation: 180,
            dx: 0,
            dy: 1
        },
        {
            id: 'east',
            rotation: 90,
            dx: 1,
            dy: 0
        },
        {
            id: 'west',
            rotation: 270,
            dx: -1,
            dy: 0
        }
    ]

    def self.from_rotation(rotation)
        find_by! rotation: (rotation % 360)
    end

    def +(other)
        self.class.from_rotation(rotation + other.rotation)
    end

    def -(other)
        self.class.from_rotation(rotation + (-other.rotation))
    end

    def -@
        self.class.from_rotation(rotation + 180)
    end
end

File: app/javascript/models/Orientation.ts
import _ from "lodash"

export enum OrientationId {
    NORTH = 'north',
    EAST = 'east',
    SOUTH = 'south',
    WEST = 'west',
}

export interface Orientation {
    id: OrientationId
}

export const ORIENTATION_NORTH: Orientation = { id: OrientationId.NORTH }
export const ORIENTATION_EAST: Orientation = { id: OrientationId.EAST }
export const ORIENTATION_SOUTH: Orientation = { id: OrientationId.SOUTH }
export const ORIENTATION_WEST: Orientation = { id: OrientationId.WEST }

const NEXT_ORIENTATION_ID_BY_ORIENTATION_ID = {
    [OrientationId.NORTH]: OrientationId.EAST,
    [OrientationId.EAST]: OrientationId.SOUTH,
    [OrientationId.SOUTH]:OrientationId.WEST,
    [OrientationId.WEST]: OrientationId.NORTH,
}

const PREVIOUS_ORIENTATION_ID_BY_ORIENTATION_ID = {
    [OrientationId.NORTH]: OrientationId.WEST,
    [OrientationId.EAST]: OrientationId.NORTH,
    [OrientationId.SOUTH]: OrientationId.EAST,
    [OrientationId.WEST]: OrientationId.SOUTH,
}

export function orientationNextOrientation({ id } : Orientation) {
    return { id: NEXT_ORIENTATION_ID_BY_ORIENTATION_ID[id] }
}

export function orientationPreviousOrientation({ id }: Orientation) {
    return { id: PREVIOUS_ORIENTATION_ID_BY_ORIENTATION_ID[id]}
}

export function orientationTransform({ id }: Orientation) {
    switch (id) {
        case OrientationId.NORTH:
            return '';
        case OrientationId.EAST:
            return 'rotate(90)';
        case OrientationId.SOUTH:
            return 'rotate(180)';
        case OrientationId.WEST:
            return 'rotate(270)';
    }
}
